src/serde_arrow_offsets.erl

```diff
@@ -0,0 +1,35 @@
+-module(serde_arrow_offsets).
+-export([new/2]).
+
+-include("serde_arrow_buffer.hrl").
+
+-spec new(
+    Value :: [serde_arrow_type:erlang_type()],
+    Type :: serde_arrow_type:arrow_type()
+) ->
+    Buffer :: #buffer{}.
+new(Values, Type) ->
+    Offsets = offsets(Values, [0], 0, Type),
+    serde_arrow_buffer:new(Offsets, {s, 32}).
+
+-spec offsets(
+    Value :: [serde_arrow_type:erlang_type()],
+    Acc :: [non_neg_integer()],
+    Offset :: non_neg_integer(),
+    Type :: serde_arrow_type:arrow_type()
+) -> Offsets :: [serde_arrow_type:erlang_type()].
+offsets([Value | Rest], Acc, Offset, Type) when (Value =:= undefined) orelse (Value =:= nil) ->
+    offsets(Rest, [Offset | Acc], Offset, Type);
+offsets([Value | Rest], Acc, Offset, Type) ->
+    CurOffset = Offset + len(Value, Type),
+    offsets(Rest, [CurOffset | Acc], CurOffset, Type);
+offsets([], Acc, _Offset, _Type) ->
+    lists:reverse(Acc).
```

Comment:
I've got to find the offsets of the list of binaries here. The offsets are used to find the position/slot of each element in the binary. This is not known as the length of a binary varies. For example, the offsets of:

```erlang
[<<1, 2>>, <<3>]
```

will be:

```
[0, 2, 3]
```

Thus the length of element 1 is 2 b (2 - 0 = 2) and it starts 0 bytes from the beginning of the buffer and ends at 2 bytes. The length of element 2 is 1 b (3 - 2 = 1) and it starts 2 bytes from the beginning of the buffer and ends at 3 bytes.

Now, the question is if the way I'm computing this appropriate. There are many ways of approaching this problem:

One is to write the a fully tail recursive function as above, accumulate in the reverse order and then reverse on return. 

Another is to write a tail recursive function and accumulate in order. Since we append in this case, a lot of data is getting copied, which on consideration makes the approach plain stupid.

Another is to write a body recursive function so that I don't need to accumulate or reverse. However, this function has to work with 2^32 elements. Could I hit a stackoverflow error if get near to that number of calls on stack?

Reply:
Current version is great. Body recursion would also work, you won't get a stackoverflow error, and sometimes it is faster too. Go with whatever you feel more readable.

Reply:
I think the tail-recursive version is better. You can't compute current offsets body-recursively, only iteratively. So when I do a body-recursive func, the offsets list would by computed body recursively, which the current offsets is computed iteratively, which makes for a weird and difficult to read function.

Reply:
For completeness, a body recursive version is imlpemented like this:

```erlang
offsets([Value | Rest], Offset, Type) when (Value =:= undefined) orelse (Value =:= nil) ->
    [Offset | offsets(Rest, Offset, Type)];
offsets([Value | Rest], Offset, Type) ->
    CurOffset = Offset + len(Value, Type),
    [CurOffset | offsets(Rest, CurOffset, Type)];
offsets([], _Offset, _Type) ->
    [].
```

so you can compute the offset in a body recursive one. :)